src/lib.rs

```diff
@@ -75,18 +78,85 @@ pub use matchers::*;
 pub type Matcher = fn(buf: &[u8]) -> bool;
 
 /// Generic information for a type
-#[derive(Debug, Eq, PartialEq)]
+#[derive(Copy, Clone)]
 pub struct Type {
-    /// The mime
-    pub mime: String,
+    matcher_type: MatcherType,
+    mime_type: &'static str,
+    extension: &'static str,
+    matcher: WrapMatcher,
+}
+
+impl Type {
+    pub(crate) const fn new(
+        matcher_type: MatcherType,
+        mime_type: &'static str,
+        extension: &'static str,
+        matcher: WrapMatcher,
+    ) -> Self {
+        Self {
+            matcher_type,
+            mime_type,
+            extension,
+            matcher,
+        }
+    }
+
+    /// Only for tests
+    #[doc(hidden)]
```

Comment:
```suggestion
    #[cfg(test)]
```

Reply:
I believe that test internals should not be exposed in the public API. Even if documentation is hidden, the code is still available for one to use. Any change to this code might be a breaking one and would require a major release bump.

Another possibility is to move this to the test module, eventually making it `pub(crate)` if this is to be used by other modules in the crate. This would be my choice to completely separate test code from library code.

Reply:
I guess the only other way is to move every test into a submodule, then #[cfg(test)] will work.

Reply:
There are several ways to do it, I am not sure about which one is best. For instance you can make `new_for_test` a  public function (with another name perhaps) in a `test_utils` module guarded by `#[cfg(test)]` that you would use in test modules which need it. Or you can just keep this suggestion.

Whatever the solution you like i believe it makes sense to remove this test only method from the public API.

What do you think ?

Reply:
Looking at other crates there's seems to be a consensus on things behind #[doc(hidden)] not being considered part of the stable api guarantees so it shouldn't be that big of a problem. I think I can update the doc so say it's for internal use only and not considered part of the stable api.

Reply:
O wait I think I know what you mean. Let me test something

Reply:
The issue with this is that `#[cfg(test)]` only works for submodules, so that's why I noticed other crates use `#[doc(hidden)]` to make some functions public to tests inside the `tests` folder

Reply:
Sorry I do not get your concern.
using `#[cfg(test)]` instead of `#[doc(hidden)]` just works isn't it ?
See for instance: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dcc88de1636ecf6839cd06e3af50c8ff

Sorry if I am missing something.

Reply:
`#[cfg(test)]` works only for [unit tests](https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html), like the one from your example, but won't work for [integration tests](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html)

Reply:
I see. I checked out your pull request and can reproduce, indeed.

Sorry to be pedantic here and maybe over looking at this. What we are discussing is not a very big deal but I am still convinced that we can do better here. Also I just noticed that we are exposing `WrapMatcher` to the public API, maybe we do not have to.

Give me a couple of hours this afternoon. I would like to experiment a few things and will come back to you with my findings. If I fail I guess the `#[doc(hidden)]` trick will make it anyway.

Reply:
> Sorry to be pedantic here and maybe over looking at this. What we are discussing is not a very big deal but I am still convinced that we can do better here. Also I just noticed that we are exposing WrapMatcher to the public API, maybe we do not have to.

Sure no problem. I don't think WrapMatcher is an issue, since we are not re-exporting it in lib.rs https://github.com/paolobarbolini/infer/blob/fa711bb618417a5d294a2dde3a02c6294b1f8b0d/src/lib.rs#L72

I ran `cargo doc --open` to confirm and I couldn't find it

Reply:
yeah my bad and edited my comment at the same time (this is where GH misses a realtime chat feature).

Reply:
Here is a small change that will hopefully address both needs: https://github.com/paolobarbolini/infer/compare/refactor...ririsoft:refactor

The idea here is that we rename `Type::new_for_test` to `Type::new` and make it officially public (and passing it a `Matcher` is mandatory). The previous `Type::new` const version remains private and is renamed to `new_static` (a better name is welcome if you have one).

By doing that we allow integration tests to use a public method, which sounds reasonable after all: integration test should use only public API and not internal ones. In addition each integration (or unit) tests can use a custom `Matcher` if needed.

Also we allow users to have their own instance of Type for their custom needs if any. An open question is whether we make the `Type::new_static` public too or not.

What do you think ?
Cheers (and thank you for your listening and patience)

Reply:
I like your change, I cherry picked your commit.

I don't think we'll be able to also make `Type::new_static` public, at least not without also having to expose `WrapMatcher`, which might not be as bad since it will take some time anyway before the const feature is stabilized and more months after that to be able to drop support for older compiler versions.